<a href="https://colab.research.google.com/github/arvind-choudhary-14/ML-ASSIGNMENTS/blob/master/17EJTCS016_ARVIND_ASSIGNMENT5_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Adult Dataset :


Step - 1 - Getting things ready for the project
Import the required libraries

Load the data
Discover and handle missing values (Similar to previous assignment)

Step - 2 - Data Preparation
Consider ‘income’ as the target variable (:P)
Remove unnecessary columns (For eg: ‘fnlwgt’)
Standardizing numerical data
Encoding categorical features
Train Test Split

Step - 3 - Train the model on Training Data

Consider ‘income’ as the target variable (-_-)
Identify the TASK in hand and choose the appropriate algorithm for the task.
Since task is Classification, use Logistic Regression

Step - 4 - Predict on Test Data

Step - 5 - Evaluate each Model


* Use evaluation metrics like Accuracy, Confusion Metric, Precision and Recall.
* Print classification report as shown in the figure below


In [ ]:
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn import svm
from sklearn.model_selection import KFold
from sklearn.decomposition import PCA
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier

## CSV Data load 

In [ ]:
missing_val = '?'
df = pd.read_csv(r'C:\Users\Arvind Choudhary\Downloads\adult.csv',na_values = missing_val)

In [ ]:
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18        NaN  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                NaN    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [ ]:
df.tail()

age     workclass  fnlwgt   education  educational-num  \
48837   27       Private  257302  Assoc-acdm               12   
48838   40       Private  154374     HS-grad                9   
48839   58       Private  151910     HS-grad                9   
48840   22       Private  201490     HS-grad                9   
48841   52  Self-emp-inc  287927     HS-grad                9   

           marital-status         occupation relationship   race  gender  \
48837  Married-civ-spouse       Tech-support         Wife  White  Female   
48838  Married-civ-spouse  Machine-op-inspct      Husband  White    Male   
48839             Widowed       Adm-clerical    Unmarried  White  Female   
48840       Never-married       Adm-clerical    Own-child  White    Male   
48841  Married-civ-spouse    Exec-managerial         Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
48837             0             0              38  United-States  <=50K  
48838             0             0              40  United-States   >50K  
48839             0             0              40  United-States  <=50K  
48840             0             0              20  United-States  <=50K  
48841         15024             0              40  United-States   >50K

In [ ]:
df.shape

(48842, 15)

***
* **There are 48842 Rows and 15 columns in the dataset.**
***

In [ ]:
df.describe(include="all").T

count unique                 top   freq     mean      std  \
age              48842    NaN                 NaN    NaN  38.6436  13.7105   
workclass        46043      8             Private  33906      NaN      NaN   
fnlwgt           48842    NaN                 NaN    NaN   189664   105604   
education        48842     16             HS-grad  15784      NaN      NaN   
educational-num  48842    NaN                 NaN    NaN  10.0781  2.57097   
marital-status   48842      7  Married-civ-spouse  22379      NaN      NaN   
occupation       46033     14      Prof-specialty   6172      NaN      NaN   
relationship     48842      6             Husband  19716      NaN      NaN   
race             48842      5               White  41762      NaN      NaN   
gender           48842      2                Male  32650      NaN      NaN   
capital-gain     48842    NaN                 NaN    NaN  1079.07  7452.02   
capital-loss     48842    NaN                 NaN    NaN  87.5023  403.005   
hours-per-week   48842    NaN                 NaN    NaN  40.4224  12.3914   
native-country   47985     41       United-States  43832      NaN      NaN   
income           48842      2               <=50K  37155      NaN      NaN   

                   min     25%     50%     75%         max  
age                 17      28      37      48          90  
workclass          NaN     NaN     NaN     NaN         NaN  
fnlwgt           12285  117550  178144  237642  1.4904e+06  
education          NaN     NaN     NaN     NaN         NaN  
educational-num      1       9      10      12          16  
marital-status     NaN     NaN     NaN     NaN         NaN  
occupation         NaN     NaN     NaN     NaN         NaN  
relationship       NaN     NaN     NaN     NaN         NaN  
race               NaN     NaN     NaN     NaN         NaN  
gender             NaN     NaN     NaN     NaN         NaN  
capital-gain         0       0       0       0       99999  
capital-loss         0       0       0       0        4356  
hours-per-week       1      40      40      45          99  
native-country     NaN     NaN     NaN     NaN         NaN  
income             NaN     NaN     NaN     NaN         NaN

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
age                48842 non-null int64
workclass          46043 non-null object
fnlwgt             48842 non-null int64
education          48842 non-null object
educational-num    48842 non-null int64
marital-status     48842 non-null object
occupation         46033 non-null object
relationship       48842 non-null object
race               48842 non-null object
gender             48842 non-null object
capital-gain       48842 non-null int64
capital-loss       48842 non-null int64
hours-per-week     48842 non-null int64
native-country     47985 non-null object
income             48842 non-null object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [ ]:
df.isnull().sum()

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64

***
**Observation :-**
* In above we can say there are 3 columns contain missing values i.e., 'workclass', 'occupation', 'native-country'.
* All the data containing missing value have the datatype as 'object'.
* Mostly 'workclass' and 'occupation' contain missing value.
* In above we can say that in 'workclass' there are 2799 missing values followed by 'occupation' having missing values 2809 and 'native-country' having missing values 857.
***

In [ ]:
df.dropna(inplace = True)

***
* **Using the `dropna` function from pandas we are removing all the missing values by removing the rows where missing values occurs.**
***

In [ ]:
df.isnull().sum()

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64

***
* **Now there is no missing value containg in the dataset.**
***

In [ ]:
df.drop('fnlwgt',axis=1,inplace=True)
df.head()

age  workclass     education  educational-num      marital-status  \
0   25    Private          11th                7       Never-married   
1   38    Private       HS-grad                9  Married-civ-spouse   
2   28  Local-gov    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  Some-college               10  Married-civ-spouse   
5   34    Private          10th                6       Never-married   

          occupation   relationship   race gender  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black   Male             0             0   
1    Farming-fishing        Husband  White   Male             0             0   
2    Protective-serv        Husband  White   Male             0             0   
3  Machine-op-inspct        Husband  Black   Male          7688             0   
5      Other-service  Not-in-family  White   Male             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
5              30  United-States  <=50K

***
* **Removing column fnlwgt from the dataset .**
***

##  Encoding categorical features 

In [ ]:
abc = df.select_dtypes(include=['object'])

In [ ]:
abc

workclass     education      marital-status         occupation  \
0           Private          11th       Never-married  Machine-op-inspct   
1           Private       HS-grad  Married-civ-spouse    Farming-fishing   
2         Local-gov    Assoc-acdm  Married-civ-spouse    Protective-serv   
3           Private  Some-college  Married-civ-spouse  Machine-op-inspct   
5           Private          10th       Never-married      Other-service   
...             ...           ...                 ...                ...   
48837       Private    Assoc-acdm  Married-civ-spouse       Tech-support   
48838       Private       HS-grad  Married-civ-spouse  Machine-op-inspct   
48839       Private       HS-grad             Widowed       Adm-clerical   
48840       Private       HS-grad       Never-married       Adm-clerical   
48841  Self-emp-inc       HS-grad  Married-civ-spouse    Exec-managerial   

        relationship   race  gender native-country income  
0          Own-child  Black    Male  United-States  <=50K  
1            Husband  White    Male  United-States  <=50K  
2            Husband  White    Male  United-States   >50K  
3            Husband  Black    Male  United-States   >50K  
5      Not-in-family  White    Male  United-States  <=50K  
...              ...    ...     ...            ...    ...  
48837           Wife  White  Female  United-States  <=50K  
48838        Husband  White    Male  United-States   >50K  
48839      Unmarried  White  Female  United-States  <=50K  
48840      Own-child  White    Male  United-States  <=50K  
48841           Wife  White  Female  United-States   >50K  

[45222 rows x 9 columns]

In [ ]:
encoder = OneHotEncoder(drop='first', sparse=False)
abc_df = pd.DataFrame(encoder.fit_transform(abc), columns=encoder.get_feature_names(abc.columns))

In [ ]:
abc_df

workclass_Local-gov  workclass_Private  workclass_Self-emp-inc  \
0                      0.0                1.0                     0.0   
1                      0.0                1.0                     0.0   
2                      1.0                0.0                     0.0   
3                      0.0                1.0                     0.0   
4                      0.0                1.0                     0.0   
...                    ...                ...                     ...   
45217                  0.0                1.0                     0.0   
45218                  0.0                1.0                     0.0   
45219                  0.0                1.0                     0.0   
45220                  0.0                1.0                     0.0   
45221                  0.0                0.0                     1.0   

       workclass_Self-emp-not-inc  workclass_State-gov  workclass_Without-pay  \
0                             0.0                  0.0                    0.0   
1                             0.0                  0.0                    0.0   
2                             0.0                  0.0                    0.0   
3                             0.0                  0.0                    0.0   
4                             0.0                  0.0                    0.0   
...                           ...                  ...                    ...   
45217                         0.0                  0.0                    0.0   
45218                         0.0                  0.0                    0.0   
45219                         0.0                  0.0                    0.0   
45220                         0.0                  0.0                    0.0   
45221                         0.0                  0.0                    0.0   

       education_11th  education_12th  education_1st-4th  education_5th-6th  \
0                 1.0             0.0                0.0                0.0   
1                 0.0             0.0                0.0                0.0   
2                 0.0             0.0                0.0                0.0   
3                 0.0             0.0                0.0                0.0   
4                 0.0             0.0                0.0                0.0   
...               ...             ...                ...                ...   
45217             0.0             0.0                0.0                0.0   
45218             0.0             0.0                0.0                0.0   
45219             0.0             0.0                0.0                0.0   
45220             0.0             0.0                0.0                0.0   
45221             0.0             0.0                0.0                0.0   

       ...  native-country_Puerto-Rico  native-country_Scotland  \
0      ...                         0.0                      0.0   
1      ...                         0.0                      0.0   
2      ...                         0.0                      0.0   
3      ...                         0.0                      0.0   
4      ...                         0.0                      0.0   
...    ...                         ...                      ...   
45217  ...                         0.0                      0.0   
45218  ...                         0.0                      0.0   
45219  ...                         0.0                      0.0   
45220  ...                         0.0                      0.0   
45221  ...                         0.0                      0.0   

       native-country_South  native-country_Taiwan  native-country_Thailand  \
0                       0.0                    0.0                      0.0   
1                       0.0                    0.0                      0.0   
2                       0.0                    0.0                      0.0   
3                       0.0                    0.0                      0.0   
4       

In [ ]:
xyz = df.select_dtypes(include=['int64'])

In [ ]:
xyz

age  educational-num  capital-gain  capital-loss  hours-per-week
0       25                7             0             0              40
1       38                9             0             0              50
2       28               12             0             0              40
3       44               10          7688             0              40
5       34                6             0             0              30
...    ...              ...           ...           ...             ...
48837   27               12             0             0              38
48838   40                9             0             0              40
48839   58                9             0             0              40
48840   22                9             0             0              20
48841   52                9         15024             0              40

[45222 rows x 5 columns]

In [ ]:
scaler = StandardScaler()
xyz_df = pd.DataFrame(scaler.fit_transform(xyz), columns = xyz.columns)

In [ ]:
xyz_df.head()

age  educational-num  capital-gain  capital-loss  hours-per-week
0 -1.024983        -1.221559     -0.146733      -0.21878       -0.078120
1 -0.041455        -0.438122     -0.146733      -0.21878        0.754701
2 -0.798015         0.737034     -0.146733      -0.21878       -0.078120
3  0.412481        -0.046403      0.877467      -0.21878       -0.078120
4 -0.344079        -1.613277     -0.146733      -0.21878       -0.910942

In [ ]:
new_df = pd.concat([xyz_df, abc_df], axis=1)

In [ ]:
new_df

age  educational-num  capital-gain  capital-loss  hours-per-week  \
0     -1.024983        -1.221559     -0.146733      -0.21878       -0.078120   
1     -0.041455        -0.438122     -0.146733      -0.21878        0.754701   
2     -0.798015         0.737034     -0.146733      -0.21878       -0.078120   
3      0.412481        -0.046403      0.877467      -0.21878       -0.078120   
4     -0.344079        -1.613277     -0.146733      -0.21878       -0.910942   
...         ...              ...           ...           ...             ...   
45217 -0.873671         0.737034     -0.146733      -0.21878       -0.244684   
45218  0.109857        -0.438122     -0.146733      -0.21878       -0.078120   
45219  1.471665        -0.438122     -0.146733      -0.21878       -0.078120   
45220 -1.251951        -0.438122     -0.146733      -0.21878       -1.743763   
45221  1.017729        -0.438122      1.854773      -0.21878       -0.078120   

       workclass_Local-gov  workclass_Private  workclass_Self-emp-inc  \
0                      0.0                1.0                     0.0   
1                      0.0                1.0                     0.0   
2                      1.0                0.0                     0.0   
3                      0.0                1.0                     0.0   
4                      0.0                1.0                     0.0   
...                    ...                ...                     ...   
45217                  0.0                1.0                     0.0   
45218                  0.0                1.0                     0.0   
45219                  0.0                1.0                     0.0   
45220                  0.0                1.0                     0.0   
45221                  0.0                0.0                     1.0   

       workclass_Self-emp-not-inc  workclass_State-gov  ...  \
0                             0.0                  0.0  ...   
1                             0.0                  0.0  ...   
2                             0.0                  0.0  ...   
3                             0.0                  0.0  ...   
4                             0.0                  0.0  ...   
...                           ...                  ...  ...   
45217                         0.0                  0.0  ...   
45218                         0.0                  0.0  ...   
45219                         0.0                  0.0  ...   
45220                         0.0                  0.0  ...   
45221                         0.0                  0.0  ...   

       native-country_Puerto-Rico  native-country_Scotland  \
0                             0.0                      0.0   
1                             0.0                      0.0   
2                             0.0                      0.0   
3                             0.0                      0.0   
4                             0.0                      0.0   
...                           ...                      ...   
45217                         0.0                      0.0   
45218                         0.0                      0.0   
45219                         0.0                      0.0   
45220                         0.0                      0.0   
45221                         0.0                      0.0   

       native-country_South  native-country_Taiwan  native-country_Thailand  \
0                       0.0                    0.0                      0.0   
1                       0.0                    0.0                      0.0   
2                       0.0                    0.0                      0.0   
3                       0.0                    0.0                      0.0   
4                       0.0                    0.0                      0.0   
...                     ...                    ...                      ...   
45217                   0.0                    0.0                      0.0   
45218                   0.0             

##  Train-Test-Split 

In [ ]:
Y = new_df.pop('income_>50K')
X = new_df

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,Y,test_size=.60,random_state=45)

## Training the model with the training Data 

In [ ]:
lr = LogisticRegression( max_iter=250)
lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=250,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

##  Prediction

In [ ]:
y_predict = lr.predict(X_test)
y_predict

array([1., 1., 0., ..., 0., 0., 0.])

## Evaluation of the model

In [ ]:
accuracy_score(y_test, y_predict)

0.8497825606250461

In [ ]:
confusion_matrix(y_test, y_predict)

array([[9475,  685],
       [1353, 2054]], dtype=int64)

In [ ]:
print(classification(y_test,y_predict))

              precision    recall  f1-score   support

         0.0       0.88      0.93      0.90     10160
         1.0       0.75      0.60      0.67      3407

    accuracy                           0.85     13567
   macro avg       0.81      0.77      0.79     13567
weighted avg       0.84      0.85      0.84     13567



In [ ]:
from sklearn.feature_selection import RFE
lr_temp = LogisticRegression( max_iter=350)
rfe = RFE(lr_temp, 30)
rfe = rfe.fit(X_train, y_train)
rfe

RFE(estimator=LogisticRegression(C=1.0, class_weight=None, dual=False,
                                 fit_intercept=True, intercept_scaling=1,
                                 l1_ratio=None, max_iter=250,
                                 multi_class='auto', n_jobs=None, penalty='l2',
                                 random_state=None, solver='lbfgs', tol=0.0001,
                                 verbose=0, warm_start=False),
    n_features_to_select=30, step=1, verbose=0)

In [ ]:
temp_df = pd.DataFrame({'Columns' : X_train.columns, 'Included' : rfe.support_, 'Ranking' : rfe.ranking_})
temp_df

Columns  Included  Ranking
0                              age     False       11
1                  educational-num      True        1
2                     capital-gain      True        1
3                     capital-loss     False       37
4                   hours-per-week     False       12
..                             ...       ...      ...
90         native-country_Thailand     False       35
91  native-country_Trinadad&Tobago     False       49
92    native-country_United-States     False       32
93          native-country_Vietnam      True        1
94       native-country_Yugoslavia     False       30

[95 rows x 3 columns]

In [ ]:
imp_col = X_train.columns[rfe.support_]
imp_col

Index(['educational-num', 'capital-gain', 'workclass_Without-pay',
       'education_1st-4th', 'education_5th-6th', 'education_Prof-school',
       'marital-status_Married-AF-spouse', 'marital-status_Married-civ-spouse',
       'occupation_Exec-managerial', 'occupation_Farming-fishing',
       'occupation_Handlers-cleaners', 'occupation_Other-service',
       'occupation_Priv-house-serv', 'relationship_Other-relative',
       'relationship_Own-child', 'relationship_Wife', 'gender_Male',
       'native-country_Columbia', 'native-country_Dominican-Republic',
       'native-country_Ecuador', 'native-country_El-Salvador',
       'native-country_India', 'native-country_Laos', 'native-country_Mexico',
       'native-country_Nicaragua', 'native-country_Outlying-US(Guam-USVI-etc)',
       'native-country_Peru', 'native-country_Scotland',
       'native-country_South', 'native-country_Vietnam'],
      dtype='object')

In [ ]:
X_train_new = X_train[imp_col]
X_train_new

educational-num  capital-gain  workclass_Without-pay  \
21220        -1.613277     -0.146733                    0.0   
8492          1.128753     -0.146733                    0.0   
22673        -0.046403     -0.146733                    0.0   
755           1.128753      0.166868                    0.0   
34307        -0.438122     -0.146733                    0.0   
...                ...           ...                    ...   
35702        -0.046403     -0.146733                    0.0   
26767        -0.046403     -0.146733                    0.0   
6618          1.912190     -0.146733                    0.0   
24894         1.128753     -0.146733                    0.0   
29828        -1.613277     -0.146733                    0.0   

       education_1st-4th  education_5th-6th  education_Prof-school  \
21220                0.0                0.0                    0.0   
8492                 0.0                0.0                    0.0   
22673                0.0                0.0                    0.0   
755                  0.0                0.0                    0.0   
34307                0.0                0.0                    0.0   
...                  ...                ...                    ...   
35702                0.0                0.0                    0.0   
26767                0.0                0.0                    0.0   
6618                 0.0                0.0                    1.0   
24894                0.0                0.0                    0.0   
29828                0.0                0.0                    0.0   

       marital-status_Married-AF-spouse  marital-status_Married-civ-spouse  \
21220                               0.0                                1.0   
8492                                0.0                                1.0   
22673                               0.0                                0.0   
755                                 0.0                                0.0   
34307                               0.0                                0.0   
...                                 ...                                ...   
35702                               0.0                                0.0   
26767                               0.0                                0.0   
6618                                0.0                                0.0   
24894                               0.0                                1.0   
29828                               0.0                                0.0   

       occupation_Exec-managerial  occupation_Farming-fishing  ...  \
21220                         0.0                         0.0  ...   
8492                          0.0                         0.0  ...   
22673                         0.0                         0.0  ...   
755                           0.0                         0.0  ...   
34307                         0.0                         0.0  ...   
...                           ...                         ...  ...   
35702                         0.0                         0.0  ...   
26767                         0.0                         0.0  ...   
6618                          0.0                         0.0  ...   
24894                         1.0                         0.0  ...   
29828                         0.0                         0.0  ...   

       native-country_El-Salvador  native-country_India  native-country_Laos  \
21220                         0.0                   0.0                  0.0   
8492                          0.0                   0.0                  0.0   
22673                         0.0                   0.0                  0.0   
755                           0.0                   0.0                  0.0   
34307                         0.0                   0.0                  0.0   
...                           ...                   ...                  ...   
35702                         0.0                   0.0                

### Best feature application

In [ ]:
lrn = LogisticRegression(max_iter=300)
lrn.fit(X_train_new, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=250,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

### Predicting the Model

In [ ]:
y_predict_new = lrn.predict(X_train_new)
y_predict_new

array([0., 1., 0., ..., 0., 1., 0.])

In [ ]:
accuracy_score(y_test, y_predict)

0.8497825606250461

***
* **Using best features we get 84.97% of accuracy score.**
***

In [ ]:
pca = PCA(whiten=True)
pca.fit(X)
np.cumsum(pca.explained_variance_ratio_)

array([0.17231214, 0.29928058, 0.41859915, 0.52288518, 0.62121152,
       0.66781356, 0.69754836, 0.72354624, 0.74943991, 0.77017519,
       0.78875128, 0.80669662, 0.8222957 , 0.83642948, 0.84981652,
       0.8613533 , 0.87204272, 0.88250789, 0.89196213, 0.90045472,
       0.9079432 , 0.91518055, 0.92166351, 0.9276611 , 0.93310102,
       0.93814771, 0.94258311, 0.94674013, 0.95079417, 0.95475762,
       0.95861136, 0.96208182, 0.96547486, 0.96881006, 0.9718258 ,
       0.97417841, 0.97648633, 0.97868178, 0.98059593, 0.98226676,
       0.98385806, 0.98536841, 0.98661583, 0.98783116, 0.98894152,
       0.98994237, 0.99067315, 0.99131579, 0.99191365, 0.9924472 ,
       0.99292544, 0.99335723, 0.99375246, 0.99410984, 0.99446161,
       0.99478634, 0.99509617, 0.99538628, 0.99566147, 0.99591763,
       0.99617243, 0.9964134 , 0.99664952, 0.99688091, 0.99710149,
       0.99731476, 0.99752243, 0.99770572, 0.99788872, 0.99805956,
       0.99822186, 0.99837038, 0.99851495, 0.9986505 , 0.99877

In [ ]:
pca = PCA(n_components=42,whiten=True)
X_PCA = pca.fit_transform(X)

In [ ]:
X_PCA

array([[-1.32184844,  0.0805435 ,  0.13034847, ...,  0.07629343,
        -0.23544425, -0.21487985],
       [ 0.14191915, -0.60993306,  0.17576669, ..., -0.51446846,
        -0.23787048,  0.21362155],
       [ 0.13880642,  0.74367826, -0.06938524, ...,  0.26648445,
         0.01861889, -0.25662365],
       ...,
       [ 0.14446594, -1.32453457,  0.31385217, ...,  0.18614643,
        -0.00458394,  0.06619306],
       [-1.74087891,  0.6839305 ,  0.02648179, ...,  0.01626948,
        -0.12204104, -0.15888688],
       [ 0.85528542, -0.72330209,  1.41193187, ...,  0.10494096,
         0.32559936, -0.07371958]])

## Train Test Split applying on the best feat.

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(X_PCA[:], Y, test_size=.30, random_state=25)

### Logistic Regression with PCA 

In [ ]:
logist = LogisticRegression(max_iter = 250)
logist.fit(x_train,y_train)
##Predicting the test data
y_predict_logistic = logist.predict(x_test)

In [ ]:
print('Accuracy score : {}'.format(accuracy_score(y_test,y_predict_logistic)))
print('\n')
print('Confusion matrix :')
print(confusion_matrix(y_test,y_predict_logistic))
print('\n')
print('Classification Report :')
print(classification_report(y_test,y_predict_logistic))

Accuracy score : 0.8502985184639198


Confusion matrix :
[[9497  663]
 [1368 2039]]


Classification Report :
              precision    recall  f1-score   support

         0.0       0.87      0.93      0.90     10160
         1.0       0.75      0.60      0.67      3407

    accuracy                           0.85     13567
   macro avg       0.81      0.77      0.79     13567
weighted avg       0.84      0.85      0.84     13567



## Support Vector Machine with PCA :

In [ ]:
svm = svm.SVC()
svm.fit(x_train,y_train)
##Predicting the test data
y_predict_svm = svm.predict(x_test)

In [ ]:
print('Accuracy score : {}'.format(accuracy_score(y_test,y_predict_svm)))
print('\n')
print('Confusion matrix :')
print(confusion_matrix(y_test,y_predict_svm))
print('\n')
print('Classification Report :')
print(classification_report(y_test,y_predict_svm))

Accuracy score : 0.8501511019385273


Confusion matrix :
[[9522  638]
 [1395 2012]]


Classification Report :
              precision    recall  f1-score   support

         0.0       0.87      0.94      0.90     10160
         1.0       0.76      0.59      0.66      3407

    accuracy                           0.85     13567
   macro avg       0.82      0.76      0.78     13567
weighted avg       0.84      0.85      0.84     13567

